swarm/src/lib.rs

```diff
@@ -499,10 +504,22 @@ where
             .into_iter()
             .map(|a| match p2p_addr(peer_id, a) {
                 Ok(address) => {
-                    let dial = match dial_opts.role_override() {
-                        Endpoint::Dialer => self.transport.dial(address.clone()),
-                        Endpoint::Listener => self.transport.dial_as_listener(address.clone()),
+                    let transport_dial_opts = {
+                        let transport_port_use = dial_opts.port_use().unwrap_or_else(|| {
+                            let is_listening_on_same_transport =
+                                any_with_same_transport(self.listeners(), &address);
+                            if is_listening_on_same_transport {
+                                PortUse::Reuse
+                            } else {
+                                PortUse::New
+                            }
```

Comment:
Continuation of https://github.com/libp2p/rust-libp2p/pull/4568/#discussion_r1339338166.

`any_with_same_transport` only matches on the transport type, but not the concrete interface. E.g. if I listen on `127.0.0.1` only, this would suggest `PortUse::Reuse` for a public IPv4 address, even though one can not dial out to the public internet through that interface.

This is a difficult problem and requires information beyond what is available in rust-libp2p today. Ideally one would have the operating system choose the right interface for an outgoing dial. See related https://github.com/libp2p/rust-libp2p/issues/4259.

For the sake of keeping it wrong but simple instead of wrong but complex, I suggest following the initial design proposed in https://github.com/libp2p/rust-libp2p/issues/4226#issuecomment-1686008058, namely:

> How about the following:
>
> - `Swarm` tracks how many listen addresses we have
> - If we have 0 listen addresses, we allocate a new port by default
> - If we have at least one listen address, we reuse an existing port by default
> - As an optimisation for the future: If we already have a connection for this 5 tuple, we allocate a new port by default

In other words to only use a new port when there is not a single listener. Though, even then, we might as well pass a `Reuse` to the transport such that it can reuse an outgoing dial port in a later `listen_on` call.

Let me know if you want me to spend more time writing this out.

Reply:
> > How about the following:
> > 
> > * `Swarm` tracks how many listen addresses we have
> > * If we have 0 listen addresses, we allocate a new port by default
> > * If we have at least one listen address, we reuse an existing port by default
> > * As an optimisation for the future: If we already have a connection for this 5 tuple, we allocate a new port by default
> 
> In other words to only use a new port when there is not a single listener. Though, even then, we might as well pass a `Reuse` to the transport such that it can reuse an outgoing dial port in a later `listen_on` call.
> 
> Let me know if you want me to spend more time writing this out.

That is what we are doing, we just split it by the transport technology. That isn't any more wrong than operating on all listen addresses, right?

If we don't split by transport technology, I think we will run into funky errors. For example:

- User listens on TCP transport
- User dials a QUIC connection
- Dial fails because we cannot reuse a port because we are not listening on one

I don't want to implement hacks in every transport where we cannot take the passed-in parameters "literally". If `libp2p-swarm` demands to reuse a port, I think we should fail if we don't have one to reuse. Otherwise, `libp2p-swarm` can never work out, how to correctly do address translation. I want to stress that I don't want to implement address translation as part of this. We should keep the scope small but I want to design under the assumption that we will uplift address translation into `libp2p-swarm`. This design requires us that transports "adhere" to the passed in parameters and fail if they can't fulfill it.

Unless I am missing something, as a bare minimum, this requires us to separate the accounting of how many listen addresses we have for a given transport. Happy to be convinced otherwise!